### PR TITLE
HoughCircles rewritten (PR #7434 updated)

### PR DESCRIPTION
**Pull request by savuor**
### This pullrequest changes

Old pull request #7434 updated:
* merge conflicts fixed
* got rid of outdated cv::Seq containters and SSE2 intrinsics
* performance improved (from 2x to 8x depending on a number of candidate circles)
* radius search partially rewritten (histogram of radiuses is build instead of sorting; it is up to 2x faster for cases with big number of edge points)

### Perf tests results

Time of 8 runs is measured in milliseconds

Canny threshold | Accumulator threshold | Original #detected circles |  New #circles | Original code | Improved Code
--------------------- | ---------------------------- | ---------------------- | ---------------- | ---------------- | --------------------
20  |  10  |  59  |  61  |  3984.81  |  1158.9
20  |  30  |  57  |  54  |  9090.84  |  2812.07
20  |  50  |  52  |  55  |  45642.2  |  6522.96
20  |  100  |  23  |  22  |  282254  |  11317.3
50  |  10  |  58  |  57  |  3057.06  |  754.356
50  |  30  |  58  |  53  |  5116.18  |  1394.3
50  |  50  |  51  |  51  |  19703.4  |  1838.52
50  |  100  |  11  |  11  |  51612.8  |  1582.63
100  |  10  |  57  |  56  |  1890.26  |  524.728
100  |  30  |  54  |  53  |  4605.53  |  681.481
100  |  50  |  34  |  31  |  9136.89  |  830.344
100  |  100  |  5  |  6  |  17737.9  |  617.138
150  |  10  |  54  |  50  |  1339.33  |  434.024
150  |  30  |  47  |  49  |  6773.36  |  468.236
150  |  50  |  13  |  14  |  3191.05  |  379.194
150  |  100  |  4  |  4  |  1517.61  |  263.796

The image was used for perf tests:
detect_circles_soda
Several other images showed even worse performance of original code.

### CODE REVIEW

File: modules/imgproc/src/hough.cpp [hunk 1]
```diff
@@ -882,6 +881,700 @@ void cv::HoughLinesP(InputArray _image, OutputArray _lines,
...
+
+    switch( method )
+    {
+    case CV_HOUGH_GRADIENT:
+        HoughCirclesGradient(_image, _circles, (float)dp, (float)minDist,
+                             minRadius, maxRadius, cannyThresh,
+                             accThresh, maxCircles, kernelSize);
+        break;
+    default:
+        CV_Error( CV_StsBadArg, "Unrecognized method id. Actually only CV_HOUGH_GRADIENT is supported." );
```
Comment (alalek) [inline modules/imgproc/src/hough.cpp]: `CV_StsBadArg` - legacy C-API

File: modules/imgproc/src/hough.cpp [hunk 2]
```diff
@@ -882,6 +881,700 @@ void cv::HoughLinesP(InputArray _image, OutputArray _lines,
...
+        acols = cvCeil(edges.cols * idp), arows = cvCeil(edges.rows * idp);
+        astep = acols + 2;
+#if CV_SIMD128
+        haveSIMD = hasSIMD128();
+#endif
+    }
+
+    ~HoughCirclesAccumInvoker() {}
+
+    HoughCirclesAccumInvoker& operator=(const HoughCirclesAccumInvoker&) {return *this;}
```
Comment (alalek) [inline modules/imgproc/src/hough.cpp]: Where is it used?

File: modules/imgproc/src/hough.cpp [hunk 3]
```diff
@@ -882,6 +881,700 @@ void cv::HoughLinesP(InputArray _image, OutputArray _lines,
     Mat(lines).copyTo(_lines);
 }
 
+/****************************************************************************************\
+*                                     Circle Detection                                   *
+\****************************************************************************************/
+template<> class DataType<Point>
```
Comment (alalek) [inline modules/imgproc/src/hough.cpp]: What is the problem?
This helper is already declared [here](https://github.com/opencv/opencv/blob/3.3.1/modules/core/include/opencv2/core/types.hpp#L194) (deprecated, will be removed).
New type traits is [here](https://github.com/opencv/opencv/blob/3.3.1/modules/core/include/opencv2/core/types.hpp#L214-L217).

File: modules/imgproc/src/hough.cpp [hunk 4]
```diff
@@ -882,6 +881,700 @@ void cv::HoughLinesP(InputArray _image, OutputArray _lines,
...
+        else
+        {
+            AutoLock lock(_lock);
+            if(accum.empty())
+            {
+                accum = accumLocal;
+            }
+            else
+            {
+                add(accum, accumLocal, accum);
```
Comment (alalek) [inline modules/imgproc/src/hough.cpp]: Processing cv::add() under locked mutex is not a good thing anyway.
Consider to use thread local containers with reduction.

File: modules/imgproc/src/hough.cpp [hunk 5]
```diff
@@ -882,6 +881,657 @@ void cv::HoughLinesP(InputArray _image, OutputArray _lines,
...
+    CV_Assert(kernelSize == -1 || kernelSize == 3 || kernelSize == 5 || kernelSize == 7);
+    dp = max(dp, 1.f);
+    float idp = 1.f/dp;
+
+    Mat edges, dx, dy;
+
+    std::vector<Point> nz;
+    int numberOfThreads = 1;
+
+    int numThreads = std::max(1, getNumThreads());
```
Comment (alalek) [inline modules/imgproc/src/hough.cpp]: It is confusing to declare variables with similar generic names.

**Comment by mschoeneck** [comment]
:+1: Nice to see that you were inspired by the old pull request and good work in further development

**Pull request merged**